extensions/amp-form/0.1/form-dirtiness.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {isDisabled, isFieldDefault, isFieldEmpty} from '../../../src/form';
+
+export const DIRTINESS_INDICATOR_CLASS = 'amp-form-dirty';
+
+export class FormDirtiness {
+  /**
+   * @param {!HTMLFormElement} form
+   */
+  constructor(form) {
+    /** @private @const {!HTMLFormElement} */
+    this.form_ = form;
+
+    /** @private {number} */
+    this.dirtyFieldCount_ = 0;
+
+    /** @private {!Object<string, boolean>} */
+    this.isFieldNameDirty_ = {};
+
+    /** @private {boolean} */
+    this.isSubmitting_ = false;
+
+    this.installEventListeners_();
```

Comment:
Will rename "listener" to "handler". But prefer not to be _that_ explicit since I have a prototype listening to other events for `amp-bind` changes as well.